openedx/core/djangoapps/content_staging/models.py

```diff
@@ -0,0 +1,125 @@
+"""
+Models for content staging (and clipboard)
+"""
+import logging
+
+from django.contrib.auth import get_user_model
+from django.core.exceptions import ValidationError
+from django.db import models
+from django.utils.translation import gettext_lazy as _
+from opaque_keys.edx.django.models import UsageKeyField
+from opaque_keys.edx.keys import LearningContextKey
+
+from openedx.core.djangoapps.content.course_overviews.api import get_course_overview_or_none
+
+log = logging.getLogger(__name__)
+
+User = get_user_model()
+
+
+class StagedContent(models.Model):
+    """
+    Each StagedContent instance represents a "piece" of content (e.g. a single
+    XBlock, or a single Unit, or a single Subsection of a course) that is not
+    currently part of any course or library, but which is available to be copied
+    into a course or library.
+
+    Use as a clipboard: for any given user, the most recent row with
+    purpose=CLIPBOARD is the "current" clipboard content. But it can only be
+    pasted if its status is READY.
+    """
+
+    class Meta:
+        verbose_name_plural = _("Staged Content")
+
+    class Purpose(models.TextChoices):
+        """ The purpose of this staged content. """
+        CLIPBOARD = "clipboard", _("Clipboard")
+
+    class Status(models.TextChoices):
+        """ The status of this staged content. """
+        # LOADING: We are actively (asynchronously) writing the OLX and related data into the staging area.
+        # It is not ready to be read.
+        LOADING = "loading", _("Loading")
+        # READY: The content is staged and ready to be read.
+        READY = "ready", _("Ready")
+        # The content has expired and this row can be deleted, along with any associated data.
+        EXPIRED = "expired", _("Expired")
+        # ERROR: The content could not be staged.
+        ERROR = "error", _("Error")
+
+    id = models.AutoField(primary_key=True)
+    # The user that created and owns this staged content. Only this user can read it.
+    user = models.ForeignKey(User, null=False, on_delete=models.CASCADE)
```

Comment:
Not something for this PR, but I'm starting to wonder if StagedContent might be better as a lower-level primitive that doesn't even necessarily have a REST API or  user information–because it's not exposed directly and relies on higher level constructs for things like permissions. For instance, I can imagine things like Unit templates (e.g. "Case Study": html, video, question) that are StagedContent but don't belong to a particular user.

In that situation StagedContent isn't responsible for figuring how who owns content or how it should be used. But only storing a consistent little set of content + assets together.